### PR TITLE
Support handshake phase in Fluentd Forward Protocol v1

### PR DESCRIPTION
**Pull request by okkez**

### CODE REVIEW

**Comment by okkez** [comment]
TLS is not supported in this PR, for now.

File: influent-transport/src/main/java/influent/internal/util/Inet4Network.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  @Override
+  public boolean contains(InetAddress address) {
+    try {
+      return network.equals(InetNetwork.maskIP(address, netmask));
+    } catch (IllegalArgumentException ex) {
+      return false;
+    }
+  }
+
+  private InetAddress generateNetmask(int mask) {
```
Comment (okumin) [inline influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: How about using [java.util.BitSet](https://docs.oracle.com/javase/jp/8/docs/api/java/util/BitSet.html)?
It may make the logic simple.

```java
// prepare InetNetwork
BitSet subnetmask = new BitSet(32);
subnetmask.set(0, 24);
BitSet networkAddress = BitSet.valueOf(InetAddress.getByName("192.168.1.0").getAddress());
networkAddress.and(subnetmask);

// test client address
BitSet peerAddress = BitSet.valueOf(InetAddress.getByName("192.168.1.1").getAddress());
peerAddress.and(subnetmask);
peerAddress.equals(networkAddress)
```

File: influent-java/src/main/java/influent/forward/NioForwardConnection.java [hunk 2]
```diff
@@ -93,15 +116,38 @@
...
     );
+    this.keepAliveEnabled = keepAliveEnabled;
 
-    channel.register(eventLoop, SelectionKey.OP_READ, this);
+    if (this.security.isEnabled()) {
+      state = ConnectionState.INIT;
+      try {
+        // SecureRandom secureRandom = SecureRandom.getInstanceStrong();
+        // Above secureRandom may block...
+        SecureRandom secureRandom = SecureRandom.getInstance("NativePRNGNonBlocking");
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/NioForwardConnection.java]: SecureRandom can be reused.
And it may be good to store it to ThreadLocal, so I will examine how to do.
Would you add the comment "TODO: reuse SecureRandom"?

File: influent-java/src/main/java/influent/forward/ForwardClient.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      return this;
+    }
+
+    public ForwardClient build() {
+      if (host != null && network != null) {
+        // both of 'host' and 'network' are specified for client
+      }
+      if (host == null && network == null) {
+        // Either of 'host' and 'network' must be specified for client
+      }
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/ForwardClient.java]: I have an idea that a static factory method accepts a host or network.
This ensures that only either of them is given.

```scala
public static Builder ofHost(String host) {
  Builder builder = new Builder().host(host);
  builder.host = host;
  return host;
}

public Builder ofNetwork(String network) {
  ……
}
```

**Comment by okumin** [comment]
Thanks for implementing authentication features.
That looks good.

> TLS is not supported in this PR, for now.

Since it is hard to review big PR, I am happy if TLS version will be implemented in dedicate PR(TLS implementation is being copied from plaintext's one and its work is boring, so I will do it not if you want to really work by yourself).

**Comment by okkez** [comment]
Thank you for review and comments.
I will implement TLS version after merge this PR.

File: influent-java/src/main/java/influent/forward/NioForwardConnection.java (hunk 2 above)
Comment (okkez) [reply influent-java/src/main/java/influent/forward/NioForwardConnection.java]: OK, I will add TODO comment.

File: influent-transport/src/main/java/influent/internal/util/Inet4Network.java (hunk 1 above)
Comment (okkez) [reply influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: I've fixed using BitSet.

File: influent-java/src/main/java/influent/forward/ForwardClient.java (hunk 3 above)
Comment (okkez) [reply influent-java/src/main/java/influent/forward/ForwardClient.java]: Can I remove `Builder#host()` and `Builder#network()` ?

File: influent-transport/src/main/java/influent/internal/util/Inet4Network.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+public class Inet4Network implements InetNetwork {
+  private BitSet netmask;
+  private BitSet network;
+
+  public Inet4Network(InetAddress base, int netmask) {
+    BitSet subnetmask = new BitSet(32);
+    subnetmask.set(32 - netmask, 32);
+    // base.getAddress() returns like [192, 168, 1, 10]
+    // but BitSet.valueOf() expects [10, 1, 168, 192]
```
Comment (okumin) [inline influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: What does this means?

File: influent-transport/src/main/java/influent/internal/util/Inet4Network.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  @Override
+  public BitSet getNetmask() {
+    return netmask;
+  }
+
+  @Override
+  public boolean contains(InetAddress address) {
+    try {
+      return network.equals(InetNetwork.maskIP(address, netmask));
```
Comment (okumin) [inline influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: Excuse me, I have overlooked that `BitSet#equals` ignores sizes.
https://docs.oracle.com/javase/jp/8/docs/api/java/util/BitSet.html#equals-java.lang.Object-

Since masked IPv6 address may matches IPv4 network address, check the version by `address instanceof Inet4Address`.

File: influent-transport/src/main/java/influent/internal/util/Inet4Network.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    } catch (IllegalArgumentException ex) {
```
Comment (okumin) [inline influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: It seems that there is no path which throws an IllegalArgumentException.

File: influent-transport/src/main/java/influent/internal/util/InetNetwork.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    byte[] reversed = new byte[ip.length];
+    for (int i = 0; i < ip.length; i++) {
+      reversed[ip.length - (i + 1)] = ip[i];
+    }
+    return reversed;
+  }
+
+  boolean contains(InetAddress address);
+  BitSet getNetwork();
+  BitSet getNetmask();
```
Comment (okumin) [inline influent-transport/src/main/java/influent/internal/util/InetNetwork.java]: These getters are not used.

File: influent-transport/src/main/java/influent/internal/util/InetNetwork.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
+package influent.internal.util;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.BitSet;
+
+public interface InetNetwork {
+  static InetNetwork getBySpec(String spec) {
+    InetAddress base;
```
Comment (okumin) [inline influent-transport/src/main/java/influent/internal/util/InetNetwork.java]: Please move this declaration into `try`.

File: influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
+
+final class MsgPackPingDecoder {
+  private static final Logger logger = LoggerFactory.getLogger(MsgPackPingDecoder.class);
+
+  private final ForwardSecurity security;
+  private final Optional<ForwardClientNode> node;
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java]: I heard that we should use Optional for return values.
http://javazuki.com/articles/optional-introduction.html

I feel odd, but this is Java style……

File: influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+  public Optional<CheckPingResult> decode(ImmutableValue value) {
+    logger.debug("decoding ping {}", value);
+    String ping;
+    String selfHostname;
+    String sharedKeySalt;
+    String sharedKey = null;
+    String sharedKeyHexDigest;
+    String username;
+    String passwordDigest;
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java]: Variables can be declared when it is used at first.

```java
String ping = decodeString(……)
```

File: influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    }
+    ImmutableArrayValue arrayValue = value.asArrayValue();
+    if (arrayValue.size() != 6) {
+      error("A ping message must have 6 elements", value);
+    }
+    ping = decodeString(arrayValue.get(0));
+    if (!ping.equals(("PING"))) {
+      error("Invalid ping message", value);
+    }
+    selfHostname = decodeString(arrayValue.get(1));
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java]: I like `clientHostname` since `self` is a little ambiguous(Influent or access client).

File: influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        // FIXME Add remote address to log
+        logger.warn("Shared key mismatch: {}", selfHostname);
+        return Optional.of(CheckPingResult.failure("Shared key mismatch"));
+      }
+
+      if (security.isUserAuthEnabled()) {
+        md.reset();
+        md.update(userAuth);
+        md.update(username.getBytes());
+        md.update("userPassword".getBytes());
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java]: TODO: use user's raw password

File: influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+                            Optional<ForwardClientNode> node,
+                            byte[] nonce,
+                            byte[] userAuth) {
+    this.security = security;
+    this.node = node;
+    this.nonce = nonce;
+    this.userAuth = userAuth;
+  }
+
+  public Optional<CheckPingResult> decode(ImmutableValue value) {
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java]: This method always returns CheckPingResult or throws an Exception.
Optional can be removed.

File: influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+import org.msgpack.value.ImmutableValue;
+import org.msgpack.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
+
+final class MsgPackPingDecoder {
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java]: This class has the responsibility to test `PING` request.
So `decode` may not be suitable.

File: influent-java/src/main/java/influent/forward/NioForwardConnection.java [hunk 15]
```diff
@@ -93,15 +116,39 @@
...
         callback,
-        chunkSizeLimit
+        chunkSizeLimit,
+        security
     );
+    this.keepAliveEnabled = keepAliveEnabled;
 
-    channel.register(eventLoop, SelectionKey.OP_READ, this);
+    if (this.security.isEnabled()) {
+      state = ConnectionState.INIT;
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/NioForwardConnection.java]: `INIT` is dropped at line 144, so it can be removed from ConnectionState.

File: influent-java/src/main/java/influent/forward/NioForwardConnection.java [hunk 16]
```diff
@@ -114,6 +161,10 @@
   public void onWritable(final SelectionKey key) {
     if (sendResponses()) {
       eventLoop.disableInterestSet(key, SelectionKey.OP_WRITE);
+      if (state == ConnectionState.HELO) {
+        state = ConnectionState.PINGPONG;
+        eventLoop.enableInterestSet(key, SelectionKey.OP_READ);
+      }
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/NioForwardConnection.java]: Please add a comment `TODO: disconnect after writing failed PONG`.
The specification says that `Server MUST disconnect the connection if authentication/authorization operation fails.`.

File: influent-java/src/main/java/influent/forward/NioForwardConnection.java [hunk 17]
```diff
@@ -138,12 +189,46 @@ private boolean sendResponses() {
...
+        receivePing(key, result -> {
+          responses.enqueue(generatePong(result));
+          eventLoop.enableInterestSet(key, SelectionKey.OP_WRITE);
+          state = ConnectionState.ESTABLISHED;
+        });
+        break;
+      case ESTABLISHED:
+        receiveRequests(key);
+        break;
+    }
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/NioForwardConnection.java:178]: Good network programming!

File: influent-java/src/main/java/influent/forward/ForwardClientNode.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
+package influent.forward;
+
+import influent.internal.util.InetNetwork;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.List;
+
+public class ForwardClientNode {
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/ForwardClientNode.java:9]: I'm understanding that `node` comes from `in_forward.rb`.
But I feel that meanings of `ForwardClient` and `ForwardClientNode` are not clear……
I'm considering good names.

File: influent-transport/src/main/java/influent/internal/util/Inet4Network.java (hunk 4 above)
Comment (okkez) [reply influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: BitSet represents little endianness, InetAddress.getBytes() returns network byte order (big endianness).

See
* https://docs.oracle.com/javase/jp/8/docs/api/java/net/InetAddress.html#getAddress--
* https://docs.oracle.com/javase/jp/8/docs/api/java/util/BitSet.html#valueOf-byte:A-

File: influent-transport/src/main/java/influent/internal/util/Inet4Network.java (hunk 5 above)
Comment (okkez) [reply influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: Should `Inet6Network#contains` consider the version same as `Inet4Network#contains`?

**Comment by okkez** [comment]
Thank you for reviewing. I'm fixing according to review comments.

File: influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java (hunk 14 above)
Comment (okkez) [reply influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java]: How about `PingMessage`  and `PingMessage#check`?

File: influent-transport/src/main/java/influent/internal/util/Inet4Network.java (hunk 4 above)
Comment (okumin) [reply influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: When using BigSet#valueOf, the order of bytes follows the original one, but bits in each byte are reversed.

```scala
scala> BitSet.valueOf(Array[Byte](192.toByte, 168.toByte, 1, 10))
res15: java.util.BitSet = {6, 7, 11, 13, 15, 16, 25, 27}
```

This is
00000011-00010101-10000000-01010000

But `Array[Byte](192.toByte, 168.toByte, 1, 10)` is
11000000-00010101-10101000-00001010

So we should reverse the result of `InetAddress.getaddress` at bit level.
However, if such a operation is required, `BitSet` may not be useful……

Comment (okkez) [reply influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: Hmm, what shoud we do?

1. Revert current implementation
1. Keep current implementation
1. Implement BitSet like class to handle `InetAddress.getAddress` w/o reversing it
1. Use some library support both IPv4 and IPv6

Comment (okumin) [reply influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: I think 1 is better.

Both the previous implementation and the current one require bitwise operations.
It was my idea that BitSet makes subnet check easy, but it was wrong.

I will write property based tests and ensure that your implementation is correct, so there is no problem even if any corner cases exist.

Sorry for the  confusion.

Comment (okkez) [reply influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: OK, I will revert BitSet related code.

> Sorry for the confusion.

No, problem!

Comment (okkez) [reply influent-transport/src/main/java/influent/internal/util/Inet4Network.java]: I've reverted BitSet related code.

File: influent-transport/src/main/java/influent/internal/util/InetNetwork.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    try {
+      int netmask = Integer.parseInt(addressComponents[1]);
+      InetAddress base = InetAddress.getByName(addressComponents[0]);
+      if (base.getAddress().length == 16) {
+        return new Inet6Network(base, netmask);
+      } else {
+        return new Inet4Network(base, netmask);
+      }
+    } catch (NumberFormatException | UnknownHostException e) {
+      throw new IllegalArgumentException("Invalid network: " + spec);
```
Comment (okumin) [inline influent-transport/src/main/java/influent/internal/util/InetNetwork.java]: Most Exception can accept an Exception.
This is helpful since a user can trace causes.

```java
throw new IllegalArgumentException("message", e);
```

File: influent-transport/src/test/scala/influent/internal/util/Inet4NetworkSpec.scala [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.net.InetAddress
+
+import org.scalatest.Matchers._
+import org.scalatest.prop.TableDrivenPropertyChecks._
+import org.scalatest.WordSpec
+
+class Inet4NetworkSpec extends WordSpec {
+  "contains" should {
+    val data1 = Table(
```
Comment (okumin) [inline influent-transport/src/test/scala/influent/internal/util/Inet4NetworkSpec.scala:11]: Good helper 👍

File: influent-java/src/main/java/influent/forward/ForwardClient.java [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+    return sharedKey;
+  }
+
+  public List<String> getUsernames() {
+    return usernames;
+  }
+
+  public static class Builder {
+    private String host = null;
+    private String network = null;
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/ForwardClient.java:44]: I want to validate these arguments and create `InetAddress` or `InetNetwork` instances early.

File: influent-java/src/main/java/influent/forward/NioForwardConnection.java [hunk 22]
```diff
@@ -93,15 +116,38 @@
                        final long chunkSizeLimit,
                        final int sendBufferSize,
                        final boolean keepAliveEnabled,
```
Comment (okumin) [inline influent-java/src/main/java/influent/forward/NioForwardConnection.java]: This means TCP keepalive configuration……

Comment (okumin) [reply influent-java/src/main/java/influent/forward/NioForwardConnection.java]: Now, I cannot decide how to configure TCP keepalive and handshake's keepalive.
(I'm thinking that it is OK that TCP keepalive is always enabled and its configuration will be removed from Influent).

How about setting `keepalive` on HELO message `true` and adding TODO comment to `generateHelo` method?

**Comment by okumin** [comment]
I'm sorry I had been too busy with a family matter. I reviewed your PR again and I feel good.
After modifying some points which I commented, would you rebase the commit history?

@okkez

**Comment by okkez** [comment]
It's no problem! Thank you for reviewing again!
I will work for this PR in next week.

File: influent-transport/src/main/java/influent/internal/util/InetNetwork.java (hunk 19 above)
Comment (okkez) [reply influent-transport/src/main/java/influent/internal/util/InetNetwork.java]: Thank you for comment. I will add original exception here.

File: influent-java/src/main/java/influent/forward/ForwardClient.java (hunk 21 above)
Comment (okkez) [reply influent-java/src/main/java/influent/forward/ForwardClient.java:44]: Do you mean like following?

```java
private InetAddress host = null;
private InetNetwork network = null;
// ...
public static Builder ofHost(String host) {
  Builder builder = new Builder();
  builder.host = InetAddress.getByName(host);
  return builder;
}
public static Builder ofNetwork(String network) {
  Builder builder = new Builder();
  builder.network = InetNetwork.getBySpec(network);
  return builder;
}
```

File: influent-java/src/main/java/influent/forward/NioForwardConnection.java (hunk 22 above)
Comment (okkez) [reply influent-java/src/main/java/influent/forward/NioForwardConnection.java]: OK, I will add TODO comment and set `keepalive` on HELO message `true`.

**Comment by okumin** [comment]
CI detected a compile error.
https://travis-ci.org/okumin/influent/builds/292952357?utm_source=github_status&utm_medium=notification

**Comment by okumin** [comment]
And could you squash commits?

**Comment by okkez** [comment]
I've squashed commits and passed CI.

File: influent-java/src/main/java/influent/forward/ForwardClient.java (hunk 21 above)
Comment (okumin) [reply influent-java/src/main/java/influent/forward/ForwardClient.java:44]: Yes.
However, errors are detected before starting a server since host/network formats are validated on creating ForwardSecurity.
So there is no problem with the current implementation.

**Comment by okumin** [comment]
I think OK!
Please PR when you have additional commits.
Thanks!

**Comment by okkez** [comment]
Thank you for merging!
